Provide info about component tree to devtools

gaearon opened the pull request:
This is a work in progress that supersedes #6486 and #6488.

Unlike them, it doesn’t use an ID map and instead lets the core emit events as they happen.
(I should’ve started with this approach, it’s way simpler!)

I revert #6488 here because it is no longer necessary but can do this as a separate PR as well.

Things to do:

- [x] Update and unmount paths
- [x] Support and test server rendering

gaearon commented:
Ping @sebmarkbage

sebmarkbage commented on src/isomorphic/devtools/ReactComponentTreeDevtool.js:
```diff
@@ -0,0 +1,180 @@
+  },
+
+  onSetChildren(id, nextChildIDs) {
+    if (ReactComponentTreeDevtool.isTopLevelWrapper(id)) {
```
This is dangerous. The API contract allows setTopLevelWrapper to be mutated and doesn't say anything about the order of execution. This assumes that it is set before the children. To do this correctly, you'd have to empty the children if setTopLevelWrapper ever gets set after I guess.

I think we can avoid this whole problem though... [see next comment]

sebmarkbage commented on src/isomorphic/ReactDebugTool.js:
```diff
@@ -58,17 +58,38 @@ var ReactDebugTool = {
   },
-  onMountRootComponent(internalInstance) {
-    emitEvent('onMountRootComponent', internalInstance);
+  onSetIsTopLevelWrapper(debugID, isTopLevelWrapper) {
```
This information is an implementation detail that I'm trying to get rid of ASAP. It shouldn't be of interest to a debug tool to know this.

Can you just bypass it and emit the child of this as the conceptual root?

gaearon replied:
One problem I had is it’s hard to know whether something *is* a top level wrapper until it’s mounted and we can do the `instance._renderedComponent._currentElement === instance._currentElement.props` duck typing check.

I can add a special static field to its type so we can know to `_debugID` to `0` in `instantiateReactComponent` for it. Then `ReactDebugTool` can easily ignore any events about it as if they never happened.

sebmarkbage commented on src/isomorphic/ReactDebugTool.js:
```diff
@@ -58,17 +58,38 @@ var ReactDebugTool = {
   },
-  onMountComponent(internalInstance) {
-    emitEvent('onMountComponent', internalInstance);
+  onSetIsComposite(debugID, isComposite) {
```
This is also an implementation detail. You can render a full React app with no native nodes and just using life-cycle methods. We could implement DOM as just a series of composites. Also, the notion of composite is overloaded because there are stateless function, functional and classes.

I don't think we need to expose this. Do you have use case?

gaearon replied:
The main use case so far is ReactPerf only shows composites in the stats. Now that I think of it it probably doesn't matter because we gather their stats independently now. I'll see if I can remove it!

sebmarkbage commented on src/isomorphic/ReactDebugTool.js:
```diff
@@ -58,17 +58,38 @@ var ReactDebugTool = {
   },
-  onUnmountComponent(internalInstance) {
-    emitEvent('onUnmountComponent', internalInstance);
+  onSetChildren(debugID, childDebugIDs) {
```
We shouldn't this now but as an FYI, I been thinking that the notion of a "child" component isn't limited to children. It's not in React components so why should it be in native.

AirBnB wanted to do this `<input placeholder={<CustomComponent />} />` for example which would have one of the attributes expand into a series of "children".

Similar things can happen with layout I think. Anyway, let's fix that later when we know when it happens.

sebmarkbage commented:
Awesome. I like this a lot. Let's just see what we can do about avoid the top level wrapper and if we really need the notion of "isComposite".

gaearon commented:
@sebmarkbage I removed `isComposite` and tried my best to hide `TopLevelWrapper`. I also noticed that I was cleaning up the information about unmounted instances too early, as I would need it in ReactPerf when the batch gets closed for analysis. This is why I tweaked it to only purge information on an explicit `purgeUnmountedComponents()` call, and in the result I made server rendering work the same way, and removed the need for passing native containers. Let me know if I missed some reason why they are important, but traversing the tree representation inside the devtool seems just as good to me, and makes the code clearer. I think this is ready now!

gaearon commented on src/isomorphic/ReactDebugTool.js:
```diff
@@ -58,17 +63,40 @@ var ReactDebugTool = {
+    emitEvent('onSetDisplayName', debugID, displayName);
+  },
+  onSetIsEmpty(debugID, isEmpty) {
+    emitEvent('onSetIsEmpty', debugID, isEmpty);
```
I added this so the tree devtool would filter them out in `getChildIDs`. Alternatively I can go back to the old approach:

* Make composite check for whether the rendered component is empty, and ignore it in `onSetChildren`
* The tree devtool becomes unaware of parent-child association between composite and empty component so it can't clean it up on unmount easily
* We have to add the native container IDs back so it's cleaned up on native container unmount instead

It feels like what I have now is simpler in terms of logic (all cleanup paths are the same and are caused by unmounting) but I can bring back native container tracking if it is more desirable.

sebmarkbage replied:
Same here. I'd prefer if we could get rid of exposing this limitation because my reconciler just won't output any nodes for something that is empty. So I won't have this concept nor this limitation.

So it is exposing an implementation detail.

> The tree devtool becomes unaware of parent-child association between composite and empty component so it can't clean it up on unmount easily

You can also just not expose it at all to the debug tool. That way it won't need to be cleaned up. Empty components doesn't have children or anything at all that is exposed.

So all you have to do is ignore any call to the debug tool that passes an instance of something "empty".

gaearon commented:
Ping @sebmarkbage, I would like to get this in asap because I’m stacking other work on top of this.

sebmarkbage commented on src/isomorphic/ReactDebugTool.js:
```diff
@@ -58,17 +63,40 @@ var ReactDebugTool = {
+    emitEvent('onSetIsEmpty', debugID, isEmpty);
+  },
+  onSetChildren(debugID, childDebugIDs) {
+    if (!isTopLevelWrapper(debugID)) {
```
Can you move this check to where you call `.onSetChildren(...)`?

The reason I want these gone is that my new reconciler won't have any of these limitations but it still wants to target this same ReactDebugTool.

sebmarkbage commented:
I'll accept to avoid rebase problems but would be good to avoid exposing `onSetIsEmpty` (btw, if anything I think it is better to call them by a simple verb like `onEmpty`) and putting top level logic in the DebugTool.

gaearon commented:
I’ll see what I can do there. Thanks!
